tests/50federation/36-state.pl

```diff
@@ -201,118 +237,208 @@
-         log_if_fail "Sent message: " . $sent_event->{event_id};
+         log_if_fail "Missing events X, Y: " . $missing_event_x->{event_id} .
+            ", " . $missing_event_y->{event_id};
+         log_if_fail "Sent events B, C: " . $sent_event_b->{event_id} .
+            ", " . $sent_event_c->{event_id};
 
          Future->needs_all(
-            $inbound_server->await_request_state_ids( $room_id )
+            $outbound_client->send_event(
+               event       => $sent_event_c,
+               destination => $first_home_server,
+            ),
+
+            $inbound_server->await_request_get_missing_events( $room_id )
             ->then( sub {
                my ( $req ) = @_;
 
-               log_if_fail "Got /state_ids request";
+               my $body = $req->body_from_json;
+               log_if_fail "/get_missing_events request", $body;
 
-               my @auth_event_ids = map { $_->{event_id} } $room->current_state_events;
+               assert_deeply_eq(
+                  $body->{latest_events},
+                  [ $sent_event_c->{event_id } ],
+                  "latest_events in /get_missing_events request",
+               );
 
-               # Don't need to be exact, synapse handles failure gracefully
+               # just return X
                $req->respond_json( {
-                  pdu_ids => [ $missing_state->{event_id}, @auth_event_ids ],
-                  auth_chain_ids => [ @auth_event_ids ],
+                  events => [ $missing_event_x ],
                } );
 
                Future->done(1);
             }),
-            $inbound_server->await_request_event( $missing_state->{event_id} )
-            ->then( sub {
-               my ( $req ) = @_;
-
-               log_if_fail "Got /event/ request";
 
-               # Don't need to be exact, synapse handles failure gracefully
-               $req->respond_json( {
-                  pdus => [ $missing_state ],
-               } );
+            $inbound_server->await_request_state_ids(
+               $room_id, $missing_event_y->{event_id},
+            )->then( sub {
+               my ( $req ) = @_;
+               log_if_fail "/state_ids request";
+
+               # build a state map from the room's current state and our extra events
+               my %state = %{ $room->{current_state} };
+               foreach my $event ( $missing_state_s, $missing_state_t ) {
+                  my $k = join "\0", $event->{type}, $event->{state_key};
+                  $state{$k} = $event;
+               }
+
+               my $resp = {
+                  pdu_ids => [
+                     map { $_->{event_id} } values( %state ),
+                  ],
+                  auth_chain_ids => [
+                     # XXX I'm not really sure why we have to return our
+                     # auth_events here, when they are already in the event
+                     map { $_->[0] } @{ $missing_event_y->{auth_events} },
```

Comment:
This is meant to be the full auth chain, rather than just the auth events of the state events.

Reply:
k thanks, will update the comment slightly